docker/docker_image_dest.go

```diff
@@ -119,11 +123,13 @@ func (c *sizeCounter) Write(p []byte) (n int, err error) {
 // to any other readers for download using the supplied digest.
 // If stream.Read() at any time, ESPECIALLY at end of input, returns an error, PutBlob MUST 1) fail, and 2) delete any data stored so far.
 func (d *dockerImageDestination) PutBlob(ctx context.Context, stream io.Reader, inputInfo types.BlobInfo, cache types.BlobInfoCache, isConfig bool) (types.BlobInfo, error) {
+	d.mutex.Lock()
+	defer d.mutex.Unlock()
```

Comment:
This serializes `PutBlob` again, AFAICS; so what’s the point?

Or is it just to parallelize `dockerImageSource` but not `dockerImageDestination`? In that case the `IsThreadSafe` property might be a better fit for the `ImageSource`/`ImageDestination` interfaces instead of `ImageTransport`.

Reply:
I understand your questions as I should have made that clearer in the PR/commit messages. At the moment I focused on the `src: docker, dest:storage`. I preferred to make `IsThreadSafe()` per transport to keep things simpler but I can move it to `Source/Destination` or try to cover the vice versa case.

Reply:
I was reflecting more on your concerns and believe that you really have a point. As we have discussed in other threads in this PR, the transports are only **partially thread-safe**. An `IsThreadSafe()` method is likely to trick users into believing those transports can be used concurrently. I decided to go for an `HasThreadSafe{Get,Put}Blob()` as this is what really matters in `Copy()` for now. This should limit the focus of this PR and aid in understanding what can be done with the transports and what not. A nice side-effect is that some insufficient locks can be removed as well.